<launch>
    <arg name="show" default="false" />
    <arg name="drone_id" default="1" />
    <arg name="viz" default="true" />
    <arg name="vins" default="false" />
    <arg name="detector" default="true" />
    <arg name="detector_depth" default="false" />
    <arg name="loop" default="true" />
    <arg name="distance" default="true" />
    <arg name="bag" default="true" />
    <arg name="bag_path" default="$(env HOME)/bags/swarm_local_pc.bag" />
    <arg name="width" default="600" />
    <arg name="height" default="312" />
    <arg name="outdoor" default="false" />
    <arg name="backend_only" default="false" />
    <arg name="debug_loop_initial_only" default="false" />
    <arg name="cgraph" default="true" />
    <arg name="max_keyframe_num" default="100" />
    <arg name="debug_no_relocalization" default="false" />
    <arg name="enable_random_keyframe_deletetion" default="true" />
    <arg name="distance_measurement_outlier_threshold" default="0.3" />
    <arg name="pub_track_result" default="false" />
    <arg name="debug_save_tracked_raw" default="false" />
    <arg name="down_as_main" default="false" />
    <arg name="enable_da" default="true" />
    <arg name="debug_only_front" default="false" />
    <arg name="debug_no_rejection" default="false" />

    <arg name="output_folder" default="false" />
    <arg name="vins_config_file" default="$(env HOME)/SwarmConfig/fisheye_ptgrey_n3/fisheye_cuda.yaml"/>
    <arg name="cam_up_config_file" default="$(env HOME)/SwarmConfig/fisheye_ptgrey_n3/up.yaml"/>
    <arg name="cam_down_config_file" default="$(env HOME)/SwarmConfig/fisheye_ptgrey_n3/down.yaml"/>

    <include file="$(find swarm_localization)/launch/loop-5-pc.launch" >
        <arg name="output" value="screen" />
        <arg name="viz" value="$(arg viz)" />
        <arg name="cgraph_path" value="/root/output/graph.dot" />
        <arg name="cgraph" value="$(arg cgraph)" />
        <arg name="enable_distance" value="$(arg distance)" />
        <arg name="enable_detection" value="$(arg detector)" />
        <arg name="enable_loop" default="$(arg loop)" />
        <arg name="debug_loop_initial_only" value="$(arg debug_loop_initial_only)" />
        <arg name="pcm_enable_debug_file" value="true" />
        <arg name="max_keyframe_num" value="$(arg max_keyframe_num)" />
        <arg name="debug_no_relocalization" value="$(arg debug_no_relocalization)" />
        <arg name="enable_random_keyframe_deletetion" value="$(arg enable_random_keyframe_deletetion)" />
        <arg name="distance_measurement_outlier_threshold" value="$(arg distance_measurement_outlier_threshold)" />
        <arg name="enable_da" value="$(arg enable_da)" />
        <arg name="debug_no_rejection" default="$(arg debug_no_rejection)" />
        <arg name="self_id" value="$(arg drone_id)" />
    </include>

    <group unless="$(arg backend_only)">
        <include file="$(find swarm_detector)/launch/detector.launch"  if="$(arg detector)" >
            <arg name="output" value="log" />
            <arg name="show" value="true" />
            <arg name="down_as_main" value="$(arg down_as_main)" />
            <arg name="drone_id" value="$(arg drone_id)" />
            <arg name="detect_duration" value="0.5" />
            <arg name="extrinsic_path" value="$(arg vins_config_file)"/>
            <arg name="width" value="$(arg width)" />
            <arg name="swarm_fused" value="/swarm_drones/swarm_drone_fused_pc"/>
            <arg name="enable_rear" value="true" />
            <arg name="debug_only_front" value="$(arg debug_only_front)" />
            <arg name="cam_file" value="$(arg cam_up_config_file)" />
            <arg name="cam_file_down" value="$(arg cam_down_config_file)" />
            <arg name="debug_save_tracked_raw" value="$(arg debug_save_tracked_raw)" />
            <arg name="pub_track_result" value="$(arg pub_track_result)" />
        </include>

        <include file="$(find vins)/launch/fisheye_node.launch" if="$(arg vins)">
            <arg name="config_file" value="$(arg vins_config_file)" />
            <arg name="output" value="log" />
        </include>
        
        <include file="$(find swarm_loop)/launch/pc-outdoor-fisheye.launch" if="$(arg outdoor)">
            <arg name="output" value="screen" />
            <arg name="show" value="false" />
            <arg name="width" value="$(arg width)" />
            <arg name="height" value="$(arg height)" />
            <arg name="superpoint_model_path" value="$(find swarm_loop)/models/superpoint_v1_rtx3080_$(arg height)x$(arg width)_fp16.trt" />
            <arg name="netvlad_model_path" value="$(find swarm_loop)/models/mobilenetvlad_$(arg height)x$(arg width)_rtx3080_fp16.trt" />
            <arg name="camera_config_path" value="$(find swarm_loop)/config/front-$(arg width).yaml" />
            <arg name="self_id" value="$(arg drone_id)" />
            <arg name="vins_config_path" value="$(arg vins_config_file)" />
        </include>

        <include file="$(find swarm_loop)/launch/node-sfisheye.launch" unless="$(arg outdoor)">
            <arg name="output" value="screen" />
            <arg name="show" value="false" />
            <arg name="width" value="$(arg width)" />
            <arg name="height" value="$(arg height)" />
            <arg name="superpoint_model_path" value="$(find swarm_loop)/models/superpoint_v1_rtx3080_$(arg height)x$(arg width)_fp16.trt" />
            <arg name="netvlad_model_path" value="$(find swarm_loop)/models/mobilenetvlad_$(arg height)x$(arg width)_rtx3080_fp16.trt" />
            <arg name="camera_config_path" value="$(find swarm_loop)/config/front-$(arg width).yaml" />
            <arg name="self_id" value="$(arg drone_id)" />
            <arg name="vins_config_path" value="$(arg vins_config_file)" />
            <arg name="debug_no_rejection" default="$(arg debug_no_rejection)" />
        </include>

        <group if="$(arg comm)">
            <include file="$(find localization_proxy)/launch/uwb_comm.launch">
                <arg name="output" value="log" />
                <arg name="sf_queue_max_size" value="40" />
            </include>

            <include file="$(find inf_uwb_ros)/launch/uwb_node.launch"> 
                <arg name="self_id" default="$(arg drone_id)" />
                <arg name="output" value="log" />
                <arg name="sim_latency" value="true" />
                <arg name="latency_buffer_size" value="5" />
            </include>
            <include file="$(find swarm_pilot)/launch/swarm_pilot.launch"> 
                <arg name="drone_id" default="$(arg drone_id)" />
                <arg name="output" default="log" />
            </include>
            <node name="gen_sim" type="gen_sim.py" pkg="swarm_pilot" />
        </group>

    </group>

    <node pkg="rosbag" type="record" name="rosbag_record_local" 
        args='record -O $(arg bag_path) -a -x "/stereo/(.*)|/vins_estimator/flattened_(.*)"'
        if="$(arg bag)" />
    <param name="/use_sim_time" value="true" />
</launch>